File: tools/cartridge.py
"""Which cartridge is in `dev/`, for the checks and tools that read one.

The Python half of `tools/cartridge.mjs`, and the same reasoning: five tools
opened `dev/pokecrystal.gbc` by name, which is right for the disassembly and
wrong for everything built from it. A hack's Makefile names its own output.

**The pairing is the careful part, not the globbing.** A symbol file describes
exactly one build -- its addresses are that ROM's memory map -- and handed the
wrong ROM it does not fail, it answers. Every read comes back plausible and
wrong. So a pair is taken when the two names agree, and a pair that has to be
assumed is said out loud.

Deliberately not a second implementation of the JS one's *rules*: both prefer a
matching stem, both fall back to one-of-each with a warning, both refuse to
guess between several. Two readers of one directory that disagreed about which
cartridge is current would be worse than either.
"""
import os
import pathlib

# What a Game Boy ROM is called. `.gb` as well, because a hack targeting the
# mono machine is still a cartridge these can read.
ROM_EXTS = (".gbc", ".gb")


def find_cartridge(dev: pathlib.Path):
    """`(rom, sym, note)` — any of the first two may be None.

    `DEV_ROM` and `DEV_SYM` override the search, which is what makes a run over
    several hacks scriptable without moving files about.
    """
    forced_rom, forced_sym = os.environ.get("DEV_ROM"), os.environ.get("DEV_SYM")
    if forced_rom and forced_sym:
        return pathlib.Path(forced_rom), pathlib.Path(forced_sym), None
    if not dev.is_dir():
        return None, None, None
    roms = sorted((f for f in dev.iterdir() if f.suffix.lower() in ROM_EXTS),
                  key=lambda f: -f.stat().st_mtime)
    syms = sorted(f for f in dev.iterdir() if f.suffix.lower() == ".sym")
    if not roms or not syms:
        return None, None, None
    by_stem = {f.stem: f for f in syms}
    for rom in roms:
        if rom.stem in by_stem:
            note = None if len(roms) == len(syms) == 1 else \
                f"using {rom.name} with {rom.stem}.sym"
            return rom, by_stem[rom.stem], note
    if len(roms) == 1 and len(syms) == 1:
        return roms[0], syms[0], (
            f"{roms[0].name} and {syms[0].name} do not share a name — assuming "
            "they are the same build. A symbol file from another build answers "
            "every read, plausibly and wrongly.")
    return None, None, (
        f"cannot tell which of {len(roms)} ROM(s) goes with which of "
        f"{len(syms)} symbol file(s). Name them alike, or set DEV_ROM/DEV_SYM.")


def cartridge_paths(root: pathlib.Path):
    """The pair, or `(None, None)`. The note is printed if there is one."""
    rom, sym, note = find_cartridge(root / "dev")
    if note:
        print(f"        [cartridge] {note}")
    return rom, sym


# Whether the "cannot pair these" warning has been given. Said once rather than
# per caller: six groups ask this, and six copies of one sentence reads as six
# problems.
_WARNED = []


def has_cartridge(root: pathlib.Path) -> bool:
    """Is there a cartridge to check against at all?

    Quiet when there is simply nothing in `dev/` -- that is the ordinary state
    of a clean checkout and every caller already says it is skipping. **Loud
    when there are files it cannot pair**, because that is a cartridge somebody
    put there expecting it to be used, and a silent skip would read as the
    checks passing.
    """
    rom, sym, note = find_cartridge(root / "dev")
    if note and not rom and not _WARNED:
        _WARNED.append(note)
        print(f"        [cartridge] {note}")
    return bool(rom and sym)

File: tools/cartridge.mjs
// Which cartridge is in `dev/`, for the tools that read one.
//
// Five tools opened `dev/pokecrystal.gbc` and `dev/pokecrystal.sym` by name,
// which is right for the disassembly and wrong for everything built from it:
// a hack's Makefile names its own output, so `pokeperidot.gbc`,
// `crystal-speedchoice.gbc`, `polishedcrystal-3.0.0.gbc`. Testing one meant
// renaming two files, and renaming files to test a thing is how you end up
// testing the wrong thing.
//
// **The pairing is the part that matters, not the globbing.** A symbol file
// describes exactly one build: its addresses are that ROM's memory map, and
// handed the wrong ROM it does not fail, it answers. Every read comes back
// plausible and wrong, which is this repository's most expensive shape of
// bug and the reason `loadSlot` and the handoff both check a ROM fingerprint
// before believing a save. So a pair is taken when the two names *agree*, and
// a pair that has to be assumed is said out loud.
import { readHeader } from '../gbcore/cartridge.js';
import { engineFor } from '../titles/contract.js';
import { pickTitle } from '../titles/pick.js';
import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import { join, extname, basename } from 'node:path';

import { Symbols } from '../gen2/symbols.js';

// What a Game Boy ROM is called. `.gb` as well as `.gbc`, because a hack that
// targets the mono machine is still a cartridge this can read.
const ROM_EXTS = ['.gbc', '.gb'];

const stemOf = (f) => basename(f, extname(f));

/**
 * The ROM and symbol file to use, or null.
 *
 * `DEV_ROM` and `DEV_SYM` override the search outright, which is what makes a
 * run over several hacks scriptable without moving files about.
 */
export function findCartridge(dev) {
  // The other half of the override, and the one that catches bugs rather than
  // enabling scripts: every tool here has a second path for when there is no
  // cartridge, and on a machine with one that path never runs.
  if (process.env.DEV_NO_CARTRIDGE) return null;
  const forced = process.env.DEV_ROM && process.env.DEV_SYM;
  if (forced) {
    return { rom: process.env.DEV_ROM, sym: process.env.DEV_SYM,
             name: stemOf(process.env.DEV_ROM), note: null };
  }
  if (!existsSync(dev)) return null;
  const all = readdirSync(dev);
  const roms = all.filter((f) => ROM_EXTS.includes(extname(f).toLowerCase()));
  const syms = all.filter((f) => extname(f).toLowerCase() === '.sym');
  if (!roms.length || !syms.length) return null;

  // Newest first, so a tree with two builds in it uses the one just made
  // rather than whichever sorts first.
  const newest = (a, b) => statSync(join(dev, b)).mtimeMs
    - statSync(join(dev, a)).mtimeMs;
  roms.sort(newest);
  const paired = roms.find((r) => syms.includes(`${stemOf(r)}.sym`));
  if (paired) {
    const one = roms.length === 1 && syms.length === 1;
    return {
      rom: join(dev, paired), sym: join(dev, `${stemOf(paired)}.sym`),
      name: stemOf(paired),
      note: one ? null : `using ${paired} with ${stemOf(paired)}.sym`,
    };
  }
  // Nothing agrees. One of each is almost certainly the pair somebody meant
  // -- a file renamed by hand, most likely -- and it is still an assumption,
  // so it is made once and said. Anything else is refused: guessing here buys
  // a run of confident wrong answers, and moving one file buys certainty.
  if (roms.length === 1 && syms.length === 1) {
    return {
      rom: join(dev, roms[0]), sym: join(dev, syms[0]), name: stemOf(roms[0]),
      note: `${roms[0]} and ${syms[0]} do not share a name — assuming they are `
            + 'the same build. A symbol file from another build answers every '
            + 'read, plausibly and wrongly.',
    };
  }
  return { rom: null, sym: null, name: null,
           note: `cannot tell which of ${roms.length} ROM(s) goes with which `
                 + `of ${syms.length} symbol file(s): ${roms.join(', ')} / `
                 + `${syms.join(', ')}. Name them alike, or set DEV_ROM and `
                 + 'DEV_SYM.' };
}

/**
 * The cartridge, opened, or null with a reason printed.
 *
 * `gb` is the one thing every reader in this app asks of a Game Boy: a byte of
 * ROM. Bank 0 is mapped low and every other bank pages into 0x4000, which is
 * `GameBoy.romByte`'s arithmetic — repeated here rather than imported because
 * `gb.js` reaches for a canvas on the way in, and it was repeated in four
 * tools before it was repeated in one.
 */
/**
 * The engine numbers this cartridge would run under, and which profile said so.
 *
 * **The tools were reading every cartridge as Crystal.** `RomData` takes an
 * engine profile and they all left it out, so the stock one applied -- which
 * is right for Crystal and for every hack that kept its shape, and wrong for
 * the one that did not: `tools/dex bulbasaur` on Polished Crystal answered
 * "no readable base-stats entry" while the app, which does pick a profile,
 * read it. A tool that disagrees with the app it is meant to check is worse
 * than no tool.
 *
 * The same `pickTitle` the app uses, from the same header and symbol table,
 * so there is one answer to "which cartridge is this" and not two.
 */
export function engineOf({ rom, symbols }) {
  const title = pickTitle({ header: readHeader(rom), symbols });
  return { title, engine: engineFor(title) };
}

/**
 * The `.sav` to read, or null -- `DEV_SAV` first, then whatever is in `dev/`.
 *
 * The same escape `DEV_ROM` and `DEV_SYM` give the other two files, and it
 * was missing: a hack's save could only be tested by copying it into `dev/`
 * beside somebody's real cartridge, which is exactly the thing those
 * variables exist to avoid.
 */
export function findSave(dev) {
  if (process.env.DEV_SAV) return process.env.DEV_SAV;
  if (process.env.DEV_NO_CARTRIDGE || !dev || !existsSync(dev)) return null;
  const found = readdirSync(dev).filter((f) => f.toLowerCase().endsWith('.sav'));
  return found.length ? join(dev, found.sort()[0]) : null;
}

export function openCartridge(dev) {
  const found = findCartridge(dev);
  if (!found) {
    console.log('no cartridge in dev/ — nothing to read');
    console.log('(a ROM is not in this repository and never will be)');
    return null;
  }
  if (found.note) console.log(`[cartridge] ${found.note}`);
  if (!found.rom) return null;
  const rom = readFileSync(found.rom);
  return {
    rom,
    symbols: new Symbols(readFileSync(found.sym, 'utf8')),
    gb: { romByte: (bank, addr) => rom[bank * 0x4000 + (addr & 0x3fff)] },
    name: found.name,
    romPath: found.rom,
    symPath: found.sym,
  };
}
